### PR TITLE
[POC] StreamCompatibleBlobPutter use temp file instead of temp blob

### PR DESCRIPTION
**Pull request by trantienduchn**

### CODE REVIEW

**Comment by trantienduchn** [comment]
```
[467bcbe4830160490c3ff97ba7c0454bf9d5ee06] [ERROR] Failures: 
[467bcbe4830160490c3ff97ba7c0454bf9d5ee06] [ERROR]   RabbitMQEventDeadLettersIntegrationTest.groupIdOfFailedEventShouldBeStoredInDeadLetter:251 1 expectation failed.
[467bcbe4830160490c3ff97ba7c0454bf9d5ee06] JSON path . doesn't match.
[467bcbe4830160490c3ff97ba7c0454bf9d5ee06] Expected: iterable over ["org.apache.james.webadmin.integration.rabbitmq.RabbitMQEventDeadLettersIntegrationTest$RetryEventsListenerGroup"] in any order
[467bcbe4830160490c3ff97ba7c0454bf9d5ee06]   Actual: []
```

**Comment by trantienduchn** [comment]
weird, test this please

**Comment by trantienduchn** [comment]
test this please

**Comment by trantienduchn** [comment]
```
[467bcbe4830160490c3ff97ba7c0454bf9d5ee06] [ERROR] serverShouldBeAbletoStartConcurrently  Time elapsed: 60.301 s  <<< ERROR!
[467bcbe4830160490c3ff97ba7c0454bf9d5ee06] org.apache.james.util.concurrency.ConcurrentTestRunner$NotTerminatedException
[467bcbe4830160490c3ff97ba7c0454bf9d5ee06] 	at org.apache.james.webadmin.WebAdminUtilsTest.serverShouldBeAbletoStartConcurrently(WebAdminUtilsTest.java:38)
```

**Comment by trantienduchn** [comment]
test this please

**Comment by trantienduchn** [comment]
```
[467bcbe4830160490c3ff97ba7c0454bf9d5ee06] [ERROR] Errors: 
[467bcbe4830160490c3ff97ba7c0454bf9d5ee06] [ERROR]   JamesDeploymentValidationTest>DeploymentValidation.validateDeployment:81 » InvalidServerResponse
[467bcbe4830160490c3ff97ba7c0454bf9d5ee06] [INFO] 
[467bcbe4830160490c3ff97ba7c0454bf9d5ee06] [ERROR] Tests run: 5, Failures: 0, Errors: 1, Skipped: 2
```
test this please

**Comment by trantienduchn** [comment]
```
[467bcbe4830160490c3ff97ba7c0454bf9d5ee06] [ERROR] Errors: 
[467bcbe4830160490c3ff97ba7c0454bf9d5ee06] [ERROR]   JamesDeploymentValidationTest>DeploymentValidation.validateDeploymentWithMailsFromSmtp:95 » ConditionTimeout
[467bcbe4830160490c3ff97ba7c0454bf9d5ee06] [INFO] 
[467bcbe4830160490c3ff97ba7c0454bf9d5ee06] [ERROR] Tests run: 5, Failures: 0, Errors: 1, Skipped: 2
```

**Comment by trantienduchn** [comment]
Gatling result:
100 users
I modified `nbUsers` from `50000` to 100 and 1 hour to 10 minute in `SimulationOnMailCorpus.injectUsersInScenario(..., nbUsers)`

and changed Configuration.ScenarioDuration from 3 hours to 3 minutes but feel it didn't affect my test

### Before
before

### After
after
users

### Notes
 I watched log file and see lots of Swift Service Unavailable - 503 due to `ComputeMessageFastViewProjectionListener` handling events

**Comment by trantienduchn** [comment]
test this please

**Comment by chibenwa** [comment]
I suspect your test not to be significative: using swift, blobs smaller than a given threshold are agressively read in memory... No copy-then-update nor tempFiles.


To be perf testing your changeset, we need to be sending bigger mails...

**Comment by trantienduchn** [comment]
> I suspect your test not to be significative: using swift, blobs smaller than a given threshold are agressively read in memory... No copy-then-update nor tempFiles.

To be perf testing your changeset, we need to be sending bigger mails...

I checked the threshold to determine big blobs in `ObjectStorageBlobStore.BUFFERED_SIZE = 256 * 1024`. When mails body computed in the scenario has size `1024 * 1024`
```
class SmtpNoAuthenticationNoEncryptionBigBodyScenario {
  private def generateMessage() : String =
    myRandom.grouped(200).flatMap(_.append(Stream('\r', '\n'))).take(1024 * 1024).mkString
  ...
```

isn't it enough?

**Comment by trantienduchn** [comment]
the latest result with gatling is worse than master branch, I didn't see any error relates to BlobPutter.putAndCompute, all came from BlobPutter.putDirecty with `Enqueuer.enqueue()` is the root cause

**Comment by Arsnael** [comment]
Reat it

File: server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/BlobPutter.java [hunk 1]
```diff
@@ -38,7 +42,45 @@
...
-    Mono<BlobId> putAndComputeId(ObjectStorageBucketName bucketName, Blob initialBlob, Supplier<BlobId> blobIdSupplier);
+    default Mono<BlobId> putAndComputeId(ObjectStorageBucketName bucketName, Blob initialBlob, Supplier<BlobId> blobIdSupplier) {
+        return Mono.using(
+            () -> temporallySaving(initialBlob, blobIdSupplier),
+            tempBlob -> putTempBlob(bucketName, tempBlob)
+                .then(Mono.just(tempBlob.blobId)),
+            TempBlob::destroy);
+    }
+
+    default TempBlob temporallySaving(Blob blob, Supplier<BlobId> blobIdSupplier) throws IOException {
```
Comment (mbaechler) [inline server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/BlobPutter.java]: with this code, the file saving is synchronous, it's bad for performances I guess

Comment (trantienduchn) [reply server/blob/blob-objectstorage/src/main/java/org/apache/james/blob/objectstorage/BlobPutter.java]: The old fashion uploading is still doing it synchronously, there is one possible explanation I can think about is writing blob into file then into swift is possible slower than writing directly into swift then copy + remove the temporally (which can be blazing fast)

**Comment by openpaas-james** [comment]
Build finished. No test results found.

**Comment by trantienduchn** [comment]
Doesn't make any performance gain

**Pull request closed without merging**